### PR TITLE
Test scoping rules for pattern locals declared within a while statement

### PR DESCRIPTION
**Pull request by AlekseyTs**
Also implement more robust long term fix for #9612.
Related to #8817.

@gafter, @dotnet/roslyn-compiler Please review.

### CODE REVIEW

**Comment by gafter** [comment]
:+1:

**Comment by AlekseyTs** [comment]
Test prtest/win/dbg/unit32 please.
http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/5170/

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs [hunk 1]
```diff
@@ -67,6 +68,11 @@ internal class SourceLocalSymbol : LocalSymbol
             _locations = ImmutableArray.Create<Location>(identifierToken.GetLocation());
         }
 
+        internal Binder Binder
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: So now locals themselves own a scope that they're declared in?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs [hunk 2]
```diff
@@ -67,6 +68,11 @@ internal class SourceLocalSymbol : LocalSymbol
...
+        {
+            get { return binder; }
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: Nit: Consider making this an expression-bodied property.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: This is consistent with other properties in this file and I see no benefit in changing the style.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: Scope is owned by a Binder, each local knows this Binder (there is no change here), now consumers of the local can get to that Binder (that is new).

Comment (agocke) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: I really like this solution.

**Comment by agocke** [comment]
:+1:

**Pull request merged**